nexus-watcher/src/events/mod.rs

```diff
@@ -91,16 +95,32 @@ impl Event {
         }))
     }
 
-    pub async fn handle(self, moderation: Arc<Moderation>) -> Result<(), DynError> {
+    pub async fn handle(&self, moderation: Arc<Moderation>) -> Result<(), DynError> {
         match self.event_type {
-            EventType::Put => self.handle_put_event(moderation).await,
-            EventType::Del => self.handle_del_event().await,
-        }
+            EventType::Put => self.handle_put_event(moderation).await?,
+            EventType::Del => self.handle_del_event().await?,
+        };
+
+        self.store_event().await
+    }
+
+    /// Stores event at redis as a member of stored set.
+    /// Sorting is done by synchronization timestamp.
+    async fn store_event(&self) -> Result<(), DynError> {
+        // We use sync timestamp instead of homeserver cursor (creation timestamp)
+        // because we want to ensure that events are processed in the order they
+        // were received and they can not be inserted in between already processed events
+        let ts_ms = SystemTime::now().duration_since(UNIX_EPOCH)?.as_millis() as f64;
```

Comment:
From a different angle: what happens if a few events are processed in the same millisecond on a very fast Nexus machine? Sorting doesn't match processing order anymore.

Reply:
see above

Reply:
also milliseconds it what I see retry manager is using.